test/functional/mempool_unbroadcast.py

```diff
@@ -0,0 +1,104 @@
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test that the mempool ensures transaction delivery by periodically sending
+to peers until a GETDATA is received."""
+
+import time
+
+from test_framework.mininode import P2PTxInvStore
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+    disconnect_nodes,
+    connect_nodes,
+    create_confirmed_utxos,
+)
+
+
+class MempoolUnbroadcastTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 2
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def run_test(self):
+        self.test_broadcast()
+        self.test_txn_removal()
+
+    def test_broadcast(self):
+        self.log.info("Test that mempool reattempts delivery of locally submitted transaction")
+        node = self.nodes[0]
+
+        min_relay_fee = node.getnetworkinfo()["relayfee"]
+        utxos = create_confirmed_utxos(min_relay_fee, node, 10)
+
+        disconnect_nodes(node, 1)
+
+        self.log.info("Generate transactions that only node 0 knows about")
+
+        # generate a wallet txn
+        addr = node.getnewaddress()
+        wallet_tx_hsh = node.sendtoaddress(addr, 0.0001)
+
+        # generate a txn using sendrawtransaction
+        us0 = utxos.pop()
+        inputs = [{"txid": us0["txid"], "vout": us0["vout"]}]
+        outputs = {addr: 0.0001}
+        tx = node.createrawtransaction(inputs, outputs)
+        node.settxfee(min_relay_fee)
+        txF = node.fundrawtransaction(tx)
+        txFS = node.signrawtransactionwithwallet(txF["hex"])
+        rpc_tx_hsh = node.sendrawtransaction(txFS["hex"])
+
+        # check that second node doesn't have these two txns
+        mempool = self.nodes[1].getrawmempool()
+        assert rpc_tx_hsh not in mempool
+        assert wallet_tx_hsh not in mempool
+
+        # ensure that unbroadcast txs are persisted to mempool.dat
+        self.restart_node(0)
+
+        self.log.info("Reconnect nodes & check if they are sent to node 1")
+        connect_nodes(node, 1)
+
+        # fast forward into the future & ensure that the second node has the txns
+        node.mockscheduler(15 * 60)  # 15 min in seconds
+        self.sync_mempools(timeout=30)
+        mempool = self.nodes[1].getrawmempool()
+        assert rpc_tx_hsh in mempool
+        assert wallet_tx_hsh in mempool
+
+        self.log.info("Add another connection & ensure transactions aren't broadcast again")
+
+        conn = node.add_p2p_connection(P2PTxInvStore())
+        node.mockscheduler(15 * 60)
+        time.sleep(5)
+        assert_equal(len(conn.get_invs()), 0)
+
+    def test_txn_removal(self):
+        self.log.info("Test that transactions removed from mempool are removed from unbroadcast set")
+        node = self.nodes[0]
+        disconnect_nodes(node, 1)
+
+        # since the node doesn't have any connections, it will not receive
```

Comment:
I don't think this is true. You added a `P2PTxInvStore` in `test_broadcast()` which I think is still connected.

Reply:
good point! looks like since `P2PTxInvStore` overwrites the `on_inv` function, it doesn't send back a `GETDATA`, so the relevant piece of functionality from a connection isn't triggered.

for this PR, I added a line to `disconnect_p2ps` so the expectations make more sense (making the comment true). 

but I do think its confusing that `P2PTxInvStore` inherits from `P2PInterface` but abandons this one piece of functionality. I'm thinking as a small followup I can have `P2PTxInvStore` call through to the parent `on_inv` after it stores the txid to behave more as expected. what do you think?

Reply:
You are missing the parenthesis behind the function: `node.disconnect_p2ps()` so this line is not doing the disconnecting at the moment. So it just "accidentally" works because `P2PTxInvStore` doesn't send a `GETDATA`. I think I would expect it to send a `GETDATA` because it is part of the framework now, but in general, I think the `P2PInterface` is there so you can override the behavior to be what you need so I don't have a strong feeling about it. If it was part of the test I would say it doesn't matter but because it moves to the framework, slight preference to call the super behavior.

Reply:
🤦‍♀️ sigh. thanks for catching! 

ya I agree with the reasoning of in test vs framework. I'll do a small followup to add parens & have `P2PTxInvStore` call through super.

Reply:
fixed here: https://github.com/bitcoin/bitcoin/pull/18807/commits/97f15b487f064cf359e8042afa25a5d0fb88bf8c